### ISSUE TITLE
"Relation to vi"

### ISSUE THREAD

**Issue by VanLaser** [abo-abo/lispy#456]
I was reading the section [Relation to vi](https://github.com/abo-abo/lispy#relation-to-vi)

**Comment by noctuid** [comment]
There are several advantages over vim. For one, lispy is far more specialized (towards working with sexps/lisp) and its operations take advantage of the current context as determined by the point (open or close delimiter or region) as much as possible.

As far as switching modes goes, vim has "smart" keys that both switch to insert and perform some action or motion (e.g.`a` and `o`). Lispy has commands like this too, and because lispy is position based, it is very easy to get out of, and it often doesn't matter if you are in it because the keys you use for commands generally aren't used in special positions. You almost never need to explicitly exit lispy special as an additional action like you sometimes need to with vim normal mode. You can just use `SPC`, `RET`, `)` and other keys you would normally use to continue coding. Even without lispy, you'd need to use these keys.

As for entering special, vim pretty much just has `ESC`, but lispy has plenty of smart ways to enter special (like going to the next close paren, top-level expression, etc.).

Finally, every time you close/exit an expression with `)`, you have the option to use the special commands. Unlike in vim if you just hit `ESC` after a pause in thought, no explicit `i` is necessary to continue entering text. I think this should at least be considered a slight advantage.

In summary, lispy is much "smarter" because it allows for automatically entering and exiting special and is more specialized for lisp than general text editing. You definitely need to manually enter special a lot, but lispy absolutely has advantages over vim's modal model in other ways.

**Comment by VanLaser** [comment]
Thanks both for the answers! I'm still not convinced by the explanations, although I can bet from a practical point of view lispy *is* very effective (just that the explanations somehow don't cover that exactly - for me, and also maybe I'm just daft).  First, I don't think putting lispy specialization in lisp as an advantage to Vim-as-a-general-text-editing-tool has any bearing with the question (which is about modes/states, entering/exiting them) - a fair comparison would be in this case between lispy and a vim lisp-oriented plugin (or evil, lisp-specialized package). It's like saying, yes, maybe lispy states are harder to enter/exit, but lipsy has a lot of *other* advantages that compensate that.

Second, saying that there are more ways to enter/exit the special state with lispy - instead of just ESC in Vim, and therefore implying that that's an advantage, I can't see it at all like that: it's still constraining me to do the special stuff either when I happen to have the cursor on a right position, *or*, if not, I'm forced to put the cursor there: in any case, the text insertion or (rather unnatural) cursor movement drives the state switching, which is for me like putting the cart before the horse; ESC is enough with Vim, because I can do it whenever I feel like, and it *means* exactly that I'm switching the state.  To put it in a more picturesque way, I can't drink water when I'm thirsty, but only when I get to the top of the hill, because there's where the fountains are placed. Switching states just became dependent on the inserted text, and I have to jump from paren to paren to drink water :)

Although I'm sure I can't "see" the things properly though because I'm tackling all this rather (just) theoretically.

**Comment by noctuid** [comment]
> First, I don't think putting lispy specialization in lisp as an advantage to Vim-as-a-general-text-editing-tool has any bearing with the question

The fact that lispy is oriented towards editing sexps is not directly related to mode-switching, but the fact that uses "special" for this purpose is 100% relevant to mode-switching and to all my points about entering/exiting special.

> a fair comparison would be in this case between lispy and a vim lisp-oriented plugin (or evil, lisp-specialized package).

I don't know if it would be more fair, but I think that comparison would come out even worse for some package with a dedicated lisp editing mode that worked like vi's normal mode. Most evil and vim packages (e.g. cleverparens, lispyville, vim-sexp, etc.) just add extra lisp support to normal state (and lispy integration in the case of lispyville). The only package I'm aware of that actually adds a dedicated state for lisp editing (evil-lisp-state) is comparably *worse* in terms of keypresses. If you add context support but keep the dedicated mode idea, you just have a worse lispy.

> It's like saying, yes, maybe lispy states are harder to enter/exit, but lipsy has a lot of other advantages that compensate that.

Lispy states are *easier* to enter/exit. Worst case scenario, you use bind/use `)` like you would use `ESC` in vim (a "mode-switching" key usable from anywhere in the buffer).

> Second, saying that there are more ways to enter/exit the special state with lispy - instead of just ESC in Vim, and therefore implying that that's an advantage

It's not just that there are more ways. It's that these ways are more powerful or natural.

> it's still constraining me to do the special stuff either when I happen to have the cursor on a right position

Just because entering special can be done implicitly doesn't mean it can't be done explicitly; if you don't happen to be in special, you can use a key to explicitly switch to special like you would with vim. The point of this is again that lispy can take advantage of the current context; the only possible disadvantage is if you find the movement of the cursor to be a visual annoyance.

> ESC is enough with Vim, because I can do it whenever I feel like, and it means exactly that I'm switching the state. Of course the position of the cursor can still matter/help, but in a loose sense: it's enough to be somewhere inside the region I'm interested in, not surgically precise.

I think there are clear advantages to lispy's precision, but again I don't see the downsides.

(*Getting off track with regards to only mode-switching here* but maybe this will make it a little more clear why it's so useful that lispy relies on the context of the point)
In terms of composability, vim's model makes a lot more sense for general text editing because you have a few actions and a ton of objects/regions. Therefore it makes sense to have the operation key first and a separate namespace/keyspace/mode for objects/things/regions. For lisp, this isn't the case. You basically just have atoms, lists, and comments. You can add a couple of things to that like top-level lists and strings, but you still have very few objects. You also have a lot more specialized actions like slurping, barfing, wrapping, etc.

If I want to delete some atom in the current top-level expression in evil, I'd have to:

1. Enter normal state (`ESC`)
2. Navigate to the correct atom (either in a dumb way like with `f` or with something that puts avy overlays over all atoms, e.g. `r a <~1-2 keys to select specific atom>`)
3. Delete the atom (something like `d i a`)

Alternatively, you could use a remote text object and do something like `ESC d i r a <~1-2 keys to select some atom>`

In lispy, you don't always need the `ESC`. If you've just finished writing a `defun` and want to jump to the previous defun, jump to some specific sexp in the current defun, or jump to any defun in the file, you don't need to explicitly enter some lisp mode. You can just press one key. If you aren't in special, you'll still need to press a key to get into it.

Since lispy is focused on lisp, it doesn't need to care about a lot of the stuff vim does and can cut down on the number of keys for a lot of actions, and for the most part, it can just combine motions and text objects. Want to move to and be in position to delete/copy an atom/symbol or list? That's something like `a <avy keys>` not `i r a <avy keys>` (basically `ra` -> `a`). Inclusion of whitespace generally doesn't matter, so you usually don't need the `i` or `a` either.

So lispy can cut down the previous action by 1-3 keys (but usually 2).

Now this isn't specifically because of lispy's "mode-switching" method, but lispy's insistence on operating in "special" positions is what makes this possible. Because lispy navigation commands keep the point in special, you'll always be in a position where you can precisely use an operation. Of course, lispy has a ton of other commands bound that aren't related to composition, but most of these still take advantage of the location of the point.

Limiting the point only to useful positions reduces the number of keys required to take useful actions. I think it's a pretty natural step to go from there to getting rid of an explicit mode-switching key because you almost never need to insert letters in these useful positions.

> and I have to jump from paren to paren to drink water

I look at it this way. Lispy gives you a much greater opportunity to jump to or already be at the drinking hole in one keypress or no keypresses. If you want to take some context-related action in vim, you'll still have to navigate to the correct place, it will just be less precise and/or often take more keypresses. Even when you can't conveniently immediately jump to or select what you want, lispy's focus only on useful positions means that it is still easier to get there compared to vim (since there aren't a bunch of commands dedicated to navigating to other buffer positions).

Maybe the lispy model doesn't click with you, but I in terms of efficiency, I see only benefits compared to vim. That's not to say lispy will always be fewer keypresses for the same action, but I think it often is and that the model itself is definitely smarter. To me, practicality/efficiency is what matters, but I personally find lispy to be more intuitive than vi as well.

If you haven't already, I'd recommend looking through lispy's full reference page and watching @abo-abo's lisp editing videos. You may find some use cases where you like lispy better. I still use evil for a lot of things (and lispy also has some modifier non-special keys ofc).

**Comment by VanLaser** [comment]
First, thank you for the elaborate answer, and putting all the time into it! 

> I don't know if it would be more fair, but I think that comparison would come out even worse for some package with a dedicated lisp editing mode that worked like vi's normal mode. Most evil and vim packages (e.g. cleverparens, lispyville, vim-sexp, etc.) just add extra lisp support to normal state (and lispy integration in the case of lispyville). The only package I'm aware of that actually adds a dedicated state for lisp editing (evil-lisp-state) is comparably worse in terms of keypresses. If you add context support but keep the dedicated mode idea, you just have a worse lispy.

What I understand from this is just that there are not well written Vim/Evil packages for Lisp, not that the model is worse.

> Just because entering special can be done implicitly doesn't mean it can't be done explicitly; if you don't happen to be in special, you can use a key to explicitly switch to special like you would with vim. The point of this is again that lispy can take advantage of the current context; the only possible disadvantage is if you find the movement of the cursor to be a visual annoyance.

But then, this *is* like an additional Vim/Evil mode/state, enhanced with implicit state change. So the advantages come when using the Vim model *plus* the implicit, contextual state changes. I would agree with that, just not with the original description, of using implicit state switches *only* as being better than Vim's explicit one. Pressing `Shift-0` or `Ctrl-[` is pretty much similar.

> In terms of composability, vim's model makes a lot more sense for general text editing because you have a few actions and a ton of objects/regions. Therefore it makes sense to have the operation key first and a separate namespace/keyspace/mode for objects/things/regions. For lisp, this isn't the case. You basically just have atoms, lists, and comments. You can add a couple of things to that like top-level lists and strings, but you still have very few objects. You also have a lot more specialized actions like slurping, barfing, wrapping, etc.

I see this as related to the first issue - that there isn't a well-written, Lisp-specialized Vim/Evil package. It shouldn't matter where I am, I should be able to compose something like "add variable" to a "let", or "insert variable" to a "let", or "change (the n-th) variable" in a "let" - where the cursor is either before a "let", or inside it, or after the last one in the buffer. I mean, it's even why you have your 'targets' package, to make the cursor position less relevant, if there is a sufficient number of well-defined targets available. So something like this, specialized for Lisp, would add a lot of efficiency to the explicit state, Evil-like, model. IMHO.

But, if lispy uses both *explicit* state switching (a la Vim) *and* *implicit* state switching, I can see why it's more powerful than "just" an explicit state switching model. The post started from the description on the first page, which - for me, at least - is a bit misleading.
 
Thanks again for the answer, I will certainly study lispy more from now on! :)

**Comment by noctuid** [comment]
> What I understand from this is just that there are not well written Vim/Evil packages for Lisp, not that the model is worse.

I think the better you make the model, the closer to lispy it gets. I can't think of an alternative model that could be improved by being made more different from lispy. I think that reusing normal state is the best way to go for an evil lisp editing package that wants to remain closer to vim. 

> I would agree with that, just not with the original description, of using implicit state switches only as being better than Vim's explicit one. As an aside, pressing Shift-0 or Ctrl-[ is pretty much similar.

That's fair. I think that the readme could potentially be made more clear because state switching is definitely not always implicit in lispy. 

> It shouldn't matter where I am, I should be able to compose something like "add variable" to a "let", or "insert variable" to a "let", or "change (the n-th) variable" in a "let" - where the cursor is either before a "let", or inside it, or after the last one in the buffer.

I think it's fair to expect some commands to work even outside of special. Many do work, but the situation could be improved. If you're already at the place where you want to be, I'd say it's usually at least as efficient to use lispy (`da(` vs. `)d` or `ESC <key to add expression to let>` vs `) <that key>` or `ESC>` vs. `)>`). Lispy shines when a command can't guess what to do in an ambiguous position. For example, you lispyville's `>` can take advantage of the parent at the point to determine what to do. If the point is not at a paren, it can guess that you want to slurp from the closing paren. If this isn't what you want to do, you either need to move to the correct paren first or bind four commands instead of two for slurping/barfing. In either case, lispy takes one less keypress because there is no ambiguity in special. The lack of ambiguity generally means less keypresses. You can move in this direction in an evil package. You can repurpose word motions to move by sexp instead, make shorter keybindings for sexps in operator-pending mode, add context-sensitive commands for things like slurping, etc. Lispyville takes this approach (optionally; it doesn't add an extra state because I don't think there's much downside to just altering normal state keybindings in lisp modes). I personally find lispy's approach to be more elegant/cool, but I wouldn't say that either is bad. It does comes down to preference. Lispyville is meant to tailor to the only evil use case as well, so free to make issues if you see anything you think can be improved.

Your description of targets may be accurate, but it can often take more keys to do an operation from anywhere (I used targets in the example in the last post). There are also a lot of operations with lispy that aren't possible (or easily possible) with targets. For example, implementing an operator for convolution would be extremely convoluted. Similarly, targets isn't suited for operations that depend on the side of a thing (like lispy's slurping/barfing). Commands like those don't really make sense as operators.

Like lispy, targets still has a focus on precise positions/objects. Targets' remote text objects are basically just like using a motion and then running an action (like lispy); the main difference is the operator/operation order and point resetting. You still have to care about where the point is, it just makes things more convenient if the point isn't already where it needs to be.

**Comment by VanLaser** [comment]
Thanks again for answering, clearly you've put much more thoughts into it, and it gives me also something to think about.
